From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ROCKNIX <rocknix@rocknix.org>
Date: Sat, 07 Mar 2026 00:00:00 +0000
Subject: [PATCH] arm64: dts: rockchip: px30: add DMC completion IRQ and DDR OPP table

The mainline px30.dtsi DMC node is missing the MCU completion interrupt
required for DDR frequency switching, and has only a single 924 MHz OPP.

Add GIC_SPI 105 as the completion_irq (from Rockchip BSP px30.dtsi)
and replace the single-OPP table with BSP frequency points:

  194 MHz -  950 mV  (suspend/idle)
  328 MHz -  950 mV  (light load)
  450 MHz -  950 mV  (moderate load)
  666 MHz - 1050 mV  (normal/gaming)
  786 MHz - 1100 mV  (high bandwidth)
  924 MHz - 1125 mV  (performance / boot frequency)

The driver queries ATF at probe for actually-supported FSPs and
disables any OPPs not available in the firmware.

Signed-off-by: ROCKNIX <rocknix@rocknix.org>
---
--- a/arch/arm64/boot/dts/rockchip/px30.dtsi
+++ b/arch/arm64/boot/dts/rockchip/px30.dtsi
@@ -348,6 +348,8 @@
 
 	dmc: dmc {
 		compatible = "rockchip,px30-dmc";
+		interrupts = <GIC_SPI 105 IRQ_TYPE_LEVEL_HIGH>;
+		interrupt-names = "complete_irq";
 		devfreq-events = <&dfi>;
 		clocks = <&cru SCLK_DDRCLK>;
 		clock-names = "dmc_clk";
@@ -359,6 +361,26 @@
 
 	dmc_opp_table: dmc_opp_table {
 		compatible = "operating-points-v2";
+		opp-194000000 {
+			opp-hz = /bits/ 64 <194000000>;
+			opp-microvolt = <950000>;
+		};
+		opp-328000000 {
+			opp-hz = /bits/ 64 <328000000>;
+			opp-microvolt = <950000>;
+		};
+		opp-450000000 {
+			opp-hz = /bits/ 64 <450000000>;
+			opp-microvolt = <950000>;
+		};
+		opp-666000000 {
+			opp-hz = /bits/ 64 <666000000>;
+			opp-microvolt = <1050000>;
+		};
+		opp-786000000 {
+			opp-hz = /bits/ 64 <786000000>;
+			opp-microvolt = <1100000>;
+		};
 		opp-924000000 {
 			opp-hz = /bits/ 64 <924000000>;
 			opp-microvolt = <1125000>;
